### ISSUE TITLE
[UWP] Layout of ViewCell inside ListView breaks since 2.5.1 on first re-use

### ISSUE THREAD

**Issue by cub-bsi** [xamarin/Xamarin.Forms#2809]
### Description

With the 2.5.1 update and the included changes from [this PR](https://github.com/xamarin/Xamarin.Forms/pull/1285), the layout of ViewCell contents inside a ListView in our app become terribly corrupted. I can try creating a repro for you, but I think the problem is so fundamental that a few screenshots should suffice.

Some info on what we have built:
- UI is mostly declared via XAML
- Problematic layout is a `ListView` with `DataTemplate` of a `ViewCell` that contains a `ContentView` whose content is what should be displayed
- This content is a large `Grid` containing several smaller ones, as well as the occasional `StackLayout`
- The coloured tiles you can see in the screenshots have their values and colors constantly (every few seconds) changing, this is all done via bindings. Sometimes they do display images instead of text as well.
- The charts are from `SfChart` library. I don't expect you to fix anything wrong with that library, but I think it's pretty clear that the `SfChart` code is not at fault here. (The same issues arise when disabling the charts, but I think the charts _also_ glitching further illustrates the problem.)
- List has enough entries for some of them be "off-screen"; we also move rows around (by modifying the `ObservableCollection` serving as `ItemSource`).

Problems that occur from 2.5.1 onwards
- Text formatting glitches inside the coloured tiles.
- Text completely stops being shown inside some coloured tiles.
- Instead of images inside the tiles, weird white circles show up.
- If you push a page on top of this and pop it again, layout completely collapses, every row has all of its stuff painted on top of one another if it is even rendered at all.

I have debugged this whole phenomenon to some degree by pulling the relevant portions of Xamarin.Forms into the project as custom implementations and trying several changes.

What it comes down to:
In [these lines](https://github.com/xamarin/Xamarin.Forms/pull/1285/files#diff-06aa4ceef3002480d062f42467a37c9fR366), if I never return a cached object, the problem does not arise, but performance is horrible, understandably.
```cs
// For illustration: First way of "fixing" the problem
static Cell RealizeItemTemplate(ITemplatedItemsList<Cell> templatedItems, 
	ElementTemplate template, object context)
{
	return template.CreateContent() as Cell;
}
```
(By the way, the only case in which `index` is ever going to be `-1`, I believe, is when `object context` is `null`, which seems to regularly happen when CellControls are re-bound, but could be checked in more explicit ways surely.)

So apparently the issue is related to when a CellControl is re-bound to display a cell that had been "off-screen" and is now supposed to be displayed again. This finally made me find another fix, which has been changing [these lines](https://github.com/xamarin/Xamarin.Forms/blob/104f3385ba28425ac4c28291a48e743fe44fe2c8/Xamarin.Forms.Platform.UAP/CellControl.cs#L162) into the following:
```cs
        // For illustration: Second and slightly better way of "fixing" the problem
        void OnDataContextChanged(FrameworkElement sender, DataContextChangedEventArgs args)
        {
            if (args.NewValue is Cell)
                SetCell(args.NewValue);
            else if (args.NewValue != null)
            {
                _newValue = args.NewValue;
            }
            // Layout of ViewCells (maybe all cells, whatever) breaks if we do not actively
            // invalidate for every new value.
            Device.BeginInvokeOnMainThread(InvalidateMeasure);
        }
```
This basically defers the `SetCell` call to `MeasureOverride` in all cases (at least for lists of ViewCells). `MeasureOverride` already has logic for this because of the `!_isListViewRealized` case in the original implementation.

Sadly, even though this fix seems kind of logical, we still occasionally see corrupted layouts. Also, our `SfChart` charts inside the ViewCells start throwing `COM Exception` if we do not also `InvalidateLayout()` on the ContentView actually contained by these cells.

So I hope that you can come up with a better solution.

Thank you.

### Basic Information

- Version with issue: 2.5.1 and later
- Last known good version: 2.5.0
- IDE: VS 2017
- Platform Target Frameworks:
  - UWP:  16299, Microsoft.NETCore.UniversalWindowsPlatform 6.0.8
- Affected Devices: Windows 10 PC

### Screenshots
Please note that any concrete data is randomly generated, so the point here is never one screenshot having a "5" and the other a "13" or some such, but rather one screenshot having a number, the other one not having anything at all.

Example of expected/normal layout:
![uwp_correct](https://user-images.githubusercontent.com/37654449/40420547-0256cf42-5e89-11e8-8a1c-6e0149c00e1b.PNG)

Broken state:
![broken](https://user-images.githubusercontent.com/37654449/40420464-d8681416-5e88-11e8-818e-806d3896f336.PNG)

Broken state after pushing a page on top and popping it again:
![after_nav](https://user-images.githubusercontent.com/37654449/40420533-f7cc5cfe-5e88-11e8-9806-e3a787976497.PNG)

**Comment by PureWeen** [comment]
@cub-bsi thank you for the deep investigation on this one

1) do you think this PR will resolve your issue? https://github.com/xamarin/Xamarin.Forms/pull/2512

2) if not can you provide a small sample so I can see what you're seeing? I made up a small sample myself with a listview but I'm not able to reproduce the effect you are seeing

**Comment by cub-bsi** [comment]
Thank you. The referenced PR has indeed solved the problem as well.

However, it introduces a new one: I register an `ItemTapped` callback on the `ListView` and after applying the changes from the PR, the `ItemTappedEventArgs.Item` is always null when I click on a row.
I have traced the situation to [this method](https://github.com/paymicro/Xamarin.Forms/blob/00ef0a0670f986233b7d0cb521479008c18cbb79/Xamarin.Forms.Platform.UAP/ListViewRenderer.cs#L578) in `ListViewRenderer`:
```cs
		void OnListItemClicked(int index)
		{
			Element.NotifyRowTapped(index, cell: null);
			_itemWasClicked = true;
		}

		void OnListItemClicked(object sender, ItemClickEventArgs e)
		{
			if (e.ClickedItem != null && TemplatedItemsView?.TemplatedItems != null)
			{
				var templatedItems = TemplatedItemsView.TemplatedItems;
				var selectedItemIndex = templatedItems.GetGlobalIndexOfItem(e.ClickedItem);

				OnListItemClicked(selectedItemIndex);
			}
		}
```
Apparently in the lower function `e.ClickedItem` is now a `ViewCell` already which leads to a call of `Element.NotifyRowTapped(-1, null);` which understandably does not enable the `ListView` backend to do anything useful.
But there is an obvious solution here in passing `e.ClickedItem` as the cell, i.e. `Element.NotifyRowTapped(index, e.ClickedItem); # index is -1 but that does not matter`. Other things than tapping might also have broken, though.

Anyhow, all this might be due to me incompletely incorporating the changes from the PR or generally lacking other changes that were applied along the way. I did however upgrade to the latest stable 3.0 (3.0.0 Service Release 1) and the problem persisted.

Cheers!

**Comment by cub-bsi** [comment]
With the patch applied I also get NPEs in `ReloadData` when the _original backing collection_ is reset via `Clear()`. I’ve marked the line in this excerpt:
```cs
        void ReloadData()
        {
            var allSourceItems = new ObservableCollection<object>();
            // === NPE in next line because ItemsSource is null ===
            foreach (var item in Element.ItemsSource)
                allSourceItems.Add(item);

            // WinRT throws an exception if you set ItemsSource directly to a CVS, so bind it.
            List.DataContext = context = new CollectionViewSource
            {
                Source = allSourceItems,
                IsSourceGrouped = Element.IsGroupingEnabled
            };
        }
```

If I introduce a null-check for `Element.ItemsSource` at the start of the function, I still get an NPE – at the bottom of ` void OnCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)` in `Device.BeginInvokeOnMainThread(() => List.UpdateLayout());`.

Again this might be because I am missing other changes, but in any case it seems like doing extensive regression-testing on the PR in question would be a good idea.

**Comment by PureWeen** [comment]
@cub-bsi   Can you provide a repro by chance? I've messed with this a little bit and I'm not able to recreate what you're referring to. For me ItemTapped and Clear is working without any issues

Here's me messing around. Maybe you can edit this one to recreate your issue?

[ListViewObservableCOllection (2).zip](https://github.com/xamarin/Xamarin.Forms/files/2040544/ListViewObservableCOllection.2.zip)

**Comment by samhouts** [comment]
I am going to close this issue, as the original problem is resolved by #2512. The ItemTapped issue is now being tracked in https://github.com/xamarin/Xamarin.Forms/issues/2996. Thank you!

**Comment by cub-bsi** [comment]
Layout becomes broken again in 3.1.0 (original issue this ticket described), but items are indeed clickable now (again)! We are back to square one.
